taxcalc/behavior.py

```diff
@@ -83,6 +90,72 @@ def update_behavior(self, revisions):
         self.set_year(precall_current_year)
         self._validate_parameter_values()
 
+    def update_behavior(self, reform):
+        # check that all revisions dictionary keys are integers
+        if not isinstance(reform, dict):
+            raise ValueError('ERROR: YYYY PARAM reform is not a dictionary')
+        if not reform:
+            return  # no reform to implement
+        reform_years = sorted(list(reform.keys()))
+        for year in reform_years:
+            if not isinstance(year, int):
+                msg = 'ERROR: {} KEY {}'
+                details = 'KEY in reform is not an integer calendar year'
+                raise ValueError(msg.format(year, details))
+        # check range of remaining reform_years
+        first_reform_year = min(reform_years)
+        if first_reform_year < self.start_year:
+            msg = 'ERROR: {} YEAR reform provision in YEAR < start_year={}'
+            raise ValueError(msg.format(first_reform_year, self.start_year))
+        if first_reform_year < self.current_year:
+            msg = 'ERROR: {} YEAR reform provision in YEAR < current_year={}'
+            raise ValueError(msg.format(first_reform_year, self.current_year))
+        last_reform_year = max(reform_years)
+        if last_reform_year > self.end_year:
+            msg = 'ERROR: {} YEAR reform provision in YEAR > end_year={}'
+            raise ValueError(msg.format(last_reform_year, self.end_year))
+        # validate reform parameter names and types
+        self._validate_parameter_names_types(reform)
+
+        ########################
+        # why do we check _ignore_errors after name/type validation but not
+        # values
+        #######################
+        if not self._ignore_errors and self.behavior_errors:
+            raise ValueError(self.behavior_errors)
+
+        #####################
+        # don't need to do cpi offset
+        # # optionally apply cpi_offset to inflation_rates and re-initialize
+        # if Policy._cpi_offset_in_reform(reform):
+        #     known_years = self._apply_reform_cpi_offset(reform)
+        #     self.set_default_vals(known_years=known_years)
+        #####################
+
+        # implement the reform year by year
+        precall_current_year = self.current_year
+        reform_parameters = set()
+        for year in reform_years:
+            self.set_year(year)
+            reform_parameters.update(reform[year].keys())
+            self._update({year: reform[year]})
+        self.set_year(precall_current_year)
+        # validate reform parameter values
+        self._validate_parameter_values(reform_parameters)
+        ####################
+        # Additional Behavior specific logic
+        # check behavior_errors and _ignore_errors after values validation, too
+        ###################
+        if not self._ignore_errors and self.behavior_errors:
+            raise ValueError(self.behavior_errors)
+
```

Comment:
`Policy` doesn't check whether there are error messages after doing value validation in `_validate_parameter_values`. Why doesn't `Policy` do this?

Reply:
Checking is the responsibility of the calling script.
See, for example, the basic recipe in the Cookbook.

Reply:
Ah, I see. Thanks.